docs/general/cairo_precompiles.md

```diff
@@ -0,0 +1,170 @@
+# Cairo Precompiles
+
+Kakarot zkEVM being a Starknet appchain, it is technically possible to run Cairo
+Contracts on Kakarot. The purpose of this document is to explain the design
+behind the Cairo precompiles, which are the Cairo contracts that are deployed on
+Kakarot to provide additional functionality to the users.
+
+## Requirements
+
+From a developer / user perspective, the precompiles should be easy to interact
+with from EVM contracts. As such, we need to provide a way to call the
+precompiles from EVM contracts, and to return the results to the EVM contracts.
+This involves:
+
+- Converting the EVM calldata to Cairo inputs
+- Converting the Cairo outputs to EVM return data
+
+From a security perspective, the precompiles should _never_ produce Cairo errors
+that would cause the transaction to revert. As we cannot catch contract call
+errors in Cairo, we need to ensure that executing a `call_contract` syscall to a
+precompile will never cause the transaction to revert, meaning that:
+
+- The target Cairo addresses should always correspond to a deployed contract
+- The selector of the Cairo function being called should always be present in
+  the contract called
+- The Cairo contract called should never _panic_
+
+From these principles, we can derive the following design.
+
+## Design
+
+Interacting with the Cairo precompiles will only be done by calling specific EVM
+contracts that have been whitelisted for this purpose. This ensures that the
+Cairo precompiles are only called by contracts that have been reviewed and
+approved regarding the security concerns mentioned above.
+
+As such, the execution flow of an EVM message will be as follows:
+
+```mermaid
+flowchart TD
+    A[Solidity contract] -->|ICairo.call_contract| B
+    B{address type}
+    B -->|EVM Contract| D[Execute contract]
+    B -->|EVM Precompiles| E[Execute EVM precompile]
+    B -->|Cairo Precompile| F{Whitelisted caller code ?}
+    F --> |yes| G[Call Cairo contract]
+    F --> |no| H[Call fails]
+```
+
+## Implementation
+
+The solidity interface for the Cairo precompiles is as follows:
+
+```solidity
+// SPDX-License-Identifier: MIT
+
+pragma solidity >=0.7.0 <0.9.0;
+
+/// @notice Solidity interface that allows developers to interact with Cairo precompile methods. The precompile has one method `call_contract`.
```

Comment:
technically speaking, this method has no impact as we will just skip the bytes4 selector of the EVM `CALLDATA` to retrieve only the arguments (contractAddress, functionSelector and data)

Reply:
Yes, it's just an interface for easy interactions 🤔

Reply:
and we could also allow for `library_call` or other starknet syscalls